bin/cheribsdtest/cheribsdtest_ipc_tagstripped.c

```diff
@@ -0,0 +1,278 @@
+ * Doesn't return.
+ */
+static int pointed_to;
+static void
+ipc_test_tagsend_pointer(int *fds, size_t bufferlen)
+{
+	int * __capability pointer_tosend = &pointed_to;
+	int * __capability pointer_received;
+	ssize_t len;
+	pid_t pid;
+	int status;
+
+	/* Quick self check to make sure we start with a tagged pointer. */
+	if (!cheri_gettag(pointer_tosend))
+		cheribsdtest_failure_errx(
+		    "pointer_tosend untagged before write");
+
+	/* Self check on minimum buffer size. */
+	if (bufferlen < sizeof(pointer_tosend))
+		cheribsdtest_failure_errx("buffer too small");
+
+	pid = CHERIBSDTEST_CHECK_SYSCALL(fork());
+	if (pid == 0) {
+		/*
+		 * Child process.  Perform the write and immediately exit.
+		 * The parent will hold both pipe endpoints open.
+		 */
+		len = CHERIBSDTEST_CHECK_SYSCALL(write(fds[0],
+		    &pointer_tosend, sizeof(pointer_tosend)));
+		CHERIBSDTEST_CHECK_EQ_SIZE(len, sizeof(pointer_tosend));
+		exit(0);
+	}
+
+	/*
+	 * Parent process.  Pick up the pieces from the child, who should have
+	 * written the data into the pipe, which remains fully open as we hold
+	 * references to both endpoints.
+	 *
+	 * XXXRW: Lazy.  We assume that the pipe buffer can hold our largest
+	 * test write without blocking, which would trigger a deadlock here as
+	 * we won't read until the full write has completed.  This is true in
+	 * FreeBSD, but may not universally be true.
+	 */
+	CHERIBSDTEST_CHECK_SYSCALL(waitpid(pid, &status, 0));
+	if (!WIFEXITED(status))
+		cheribsdtest_failure_errx("waitpid !WIFEXITED");
+	if (WEXITSTATUS(status) != 0)
+		cheribsdtest_failure_errx(
+		    "child returned non-zero status (%d)",
+		    WEXITSTATUS(status));
+
+	/*
+	 *
+	 * For simplicity, assume arrives in a single read.  Should be true
+	 * in practice, but isn't really a correct assumption for IPC.
+	 */
+	len = CHERIBSDTEST_CHECK_SYSCALL(read(fds[1], &pointer_received,
+	    sizeof(pointer_received)));
+	CHERIBSDTEST_CHECK_EQ_SIZE(len, sizeof(pointer_received));
+
+	/*
+	 * Bytewise comparison of visible data.
+	 */
+	if (memcmp(&pointer_tosend, &pointer_received,
+	    sizeof(pointer_tosend)))
+		cheribsdtest_failure_errx("pointer value mismatch");
+
+	/*
+	 * Check tag.
+	 */
+	if (cheri_gettag(pointer_received))
+		cheribsdtest_failure_errx("received tagged pointer value");
+
+	close(fds[0]);
+	close(fds[1]);
+
+	/*
+	 * Tag correctly stripped by IPC transit.
+	 */
+	cheribsdtest_success();
+}
+
+/*
+ * In the following tests, 8KiB is selected because it is >= PIPE_MINDIRECT
+ * while still <= PIPSIZ.  This means that pipe VM optimizations will trigger
+ * in FreeBSD, but sockets won't overfill with default socket-buffer sizes.
+ */
+CHERIBSDTEST(test_ipc_pipe_capsize_tagstripped,
+    "check that pipe IPC strips tags for a pointer-size write")
+{
+	int fds[2];
+
+	ipc_pipe_create(fds);
+	ipc_test_tagsend_pointer(fds, sizeof(int * __capability));
+}
+
+CHERIBSDTEST(test_ipc_pipe_8k_tagstripped,
+    "check that pipe IPC strips tags for an 8k write")
```

Comment:
This does a single pointer-size write.

Reply:
Fixed by reintroducing and properly using ```buffer```.